rfc/004-priority-based-mempool.md

```diff
@@ -0,0 +1,145 @@
+
+- Junjia He (@ninjaahhh)
+
+## Context
+
+User-defined priorities for blockchain transactions can be useful, e.g. `gasPrice` in Ethereum blockchain works as a first-price auction mechanism to cope with limited capacity. However in Tendermint, current approach for packing transactions into a block follows a first-in-first-out (FIFO) style which is baked in Tendermint engine and can't be affected by ABCI applications.
+
+To make Tendermint support priority-based transaction processing, two main aspects of current architecture in Tendermint need to be updated: one is on ABCI's interface for accepting a transaction (`CheckTx`), and another is on Tendermint's mempool design and implementation. This RFC proposes corresponding changes to address these two aspects.
+
+## Proposal
+
+Add a new field `priority` to the ABCI `ResponseCheckTx` message:
+
+```proto
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7
+      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string codespace = 8;
+  uint64 priority  = 9;
+}
+```
+
+Then ABCI developers can add application-specific user-defined priorities for the transactions. This addition is naturally backward-compatible because by default all requests will be returned with 0 priority and on mempool side the process will fallback to FIFO.
+
+The next step is to make current mempool implementation priority-aware. We can have 3 phases to gradually implement this feature to make as least breaking changes as possible.
+
+### Phase 0: Keep existing CList-based mempool
+
+Goal:
+
+1. Zero cost on processing "legacy" transactions (i.e. those have 0 priority)
+2. When creating a block, transactions with higher priorities should be included earlier
+
+The first phase will be experimental and only achieves the desired functionality *with potential performance degradation*. The main purpose is to design the code flow and underlying APIs accordingly, without dramatic code changes on existing data structures / algorithms.
+
+The simplest approach I can think of is:
+
+1. When reading `CheckTx` responses, mempool keeps a single counter called `maxPriority`. If all transactions have 0 priorities, the counter should simply have the same 0 value
+2. When `Mempool.ReapMaxBytesMaxGas` is called by consensus engine, mempool would check its `maxPriority` value:
+    - if it's zero, it should follow existing logic to return eligible transactions in FIFO style
+    - if not, should return eligible transactions ranked by their priorities (either by `O(n^2)` brute-force or `O(nlogn)` sort). Then during commit, update mempool's `maxPriority` accordingly since transactions will be removed from mempool
```

Comment:
not sure how viable it is to sort mempool after every block

Reply:
It should be nlogn really. What's the typical size of the mempool?

Reply:
I think the default is 5000 txs

Reply:
The mempool size is currently a function of the software being unable to handle more. Ideally it should further scale to demand, especially as more L1 scaling solutions land. 

For ref, ETH per txstreet.com is >75k txs in mempool.

Reply:
agree O(n^2) brute force should be a no-go. will update